clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -248,4 +261,28 @@ public void updateAssignment(ConsumerGroupHeartbeatResponseData.Assignment assig
         }
         maybeTransitionToStable();
     }
+
+    @Override
+    public void completeReconcile(Set<TopicPartition> revokedPartitions,
+                                  Set<TopicPartition> assignedPartitions,
+                                  Optional<KafkaException> callbackError) {
+        if (callbackError.isPresent()) {
+            // TODO: how to react to callback errors?
+        }
+
+        assignmentReconciler.completeReconcile(revokedPartitions, assignedPartitions);
+        transitionTo(MemberState.STABLE);
+        // TODO: update state to signal the HeartbeatRequestManager to send an ACK heartbeat
+    }
+
+    @Override
+    public void completeLost(Set<TopicPartition> lostPartitions, Optional<KafkaException> callbackError) {
+        if (callbackError.isPresent()) {
+            // TODO: how to react to callback errors?
+        }
+
+        assignmentReconciler.completeLost(lostPartitions);
+        transitionTo(MemberState.UNJOINED);
+        // TODO: update state to signal the HeartbeatRequestManager to send an ACK heartbeat
```

Comment:
uhm we do need to sort this out. One option I'm thinking about is I'm aiming at being able to have states that will easily indicate to the HB manager if it "shouldHeartbeatNow" (FENCED || RECONCILIATION_COMPLETED), making sure that we only transition to those when truly ready to HB (all callbacks completed). 
Right now the HB manager has a "shouldSendHeartbeat" but that's conceptually indicating more of a `shouldSkipHeartbeat` (if FAILED or NOT_IN_GROUP) so we need to be mindful about identifying both situations in the HB manager poll. 
Thoughts?

Reply:
We just need to rejoin to reenable heartbeat right? We should'nt need to signal the HB manager explicitly.

Reply:
you're right, not here, I meant it for the reconciliation complete which has the exact same TODO and we do need to sort that out. Thoughts?

Reply:
Thinking more about this, wondering if this couple of new states would do:

- existing RECONCILING : starts when member gets new assignment and triggers callbacks. No HB.
- new RECONCILIATION_COMPLETE : starts when callbacks complete. Send HB on this state (and end the state when HB is sent out, transitioning to STABLE).
- existing FENCED (renamed to RELEASING_ASSIGNMENT): starts when member receives a fencing error, resets epoch to 0 and triggers callbacks. No HB. 
- new RELEASING_ASSIGNMENT_COMPLETE: starts when callback completes. Send HB on this state (and end state when HB is sent out, transitioning to UNJOINED)

Based on these, the HB manager could send HB based on the interval or if `shouldHeartBeatNow` true (state == RECONCILIATION_COMPLETE || RELEASING_ASSIGNMENT_COMPLETE)

Reply:
@dajac @lianetm @philipnee Please take a look at the [state machine diagram](https://cwiki.apache.org/confluence/display/KAFKA/Consumer+rebalance#Consumerrebalance-RebalanceStateMachine) on the AK wiki.